src/metabase/lib/equality.cljc

```diff
@@ -123,362 +110,188 @@
+
+  - Integer IDs in the `ref` are matched by ID; this usually is unambiguous. In the case of multiple joins on one table,
+    the `:join-alias` settles the question.
+    - There may be broken cases where the ID must be resolved to a name or `:lib/desired-column-alias` and matched.
+      `query` and `stage-number` are required for this case.
+  - For string IDs, these are checked against `:lib/desired-column-alias` first.
+    - If that doesn't match any columns, `:name` is compared next.
+    - If that *still* doesn't match, and `query` and `stage-number` were supplied.
+
+  Returns the column, or nil if no match is found."
+  ([a-ref columns]
+   (find-matching-column a-ref columns {}))
+
+  ([[ref-kind opts ref-id :as a-ref] :- ::lib.schema.ref/ref
+    columns                          :- [:sequential lib.metadata/ColumnMetadata]
+    {:keys [generous?]}              :- FindMatchingColumnOptions]
+   (case ref-kind
+     ;; Aggregations are referenced by the UUID of the column being aggregated.
+     :aggregation  (->> columns
+                        (filter #(clojure.core/= (:lib/source %) :source/aggregations))
+                        (filter #(clojure.core/= (:lib/source-uuid %) ref-id))
+                        first)
+     ;; Expressions are referenced by name; fields by ID or name.
+     (:expression
+       :field)     (let [plausible (if (string? ref-id)
+                                     (plausible-matches-for-name ref-id (:join-alias opts) columns)
+                                     (plausible-matches-for-id   ref-id (:join-alias opts) columns generous?))]
+                     (case (count plausible)
+                       0 nil
+                       1 (first plausible)
+                       (disambiguate-matches a-ref plausible)))
+     (throw (ex-info "Unknown type of ref" {:ref a-ref}))))
+
+  ([query stage-number a-ref columns]
+   (find-matching-column query stage-number a-ref columns {}))
+
+  ([query                              :- [:maybe ::lib.schema/query]
+    stage-number                       :- :int
+    [_ref-kind _opts ref-id :as a-ref] :- ::lib.schema.ref/ref
+    columns                            :- [:sequential lib.metadata/ColumnMetadata]
+    opts                               :- FindMatchingColumnOptions]
+   (or (find-matching-column a-ref columns opts)
+       (when (and query (number? ref-id))
+         (when-let [resolved (resolve-field-id query stage-number ref-id)]
+           (find-matching-column (assoc a-ref 2 (or (:lib/desired-column-alias resolved)
+                                                    (:name resolved)))
+                                 (for [col columns]
+                                   (if (:id col)
+                                     (assoc col
+                                            :name                     "____lib.equality/noname"
+                                            :lib/desired-column-alias "____lib.equality/noname")
```

Comment:
That is why. (I've documented that in the docstring and a nearby comment now.)

I think I was hesitant to remove anything because of indexes being important in the original code. But this function returns the matching column, so it's fine to `(remove :id columns)` instead.